bionic/code_analyzer.py

```diff
@@ -0,0 +1,181 @@
+        # analyzing the depth of the value instead.
+        if obj_id in self._object_depth_by_id:
+            self._hash.update(CodeHashPrefix.CIRCULAR_REF.value)
+            obj = self._object_depth_by_id[obj_id]
+            obj_id = id(obj)
+
+        self._object_depth_by_id[obj_id] = depth
+        self._update(obj, depth)
+        del self._object_depth_by_id[obj_id]
+
+    def _update(self, obj, depth):
+        """
+        Contains the logic that analyzes the objects and encodes them
+        into bytes that are added to the hash.
+        """
+        if isinstance(obj, bytes):
+            self._hash.update(CodeHashPrefix.BYTES.value)
+            self._hash.update(obj)
+            self._hash.update(get_size_in_bytes(obj))
+
+        elif isinstance(obj, bytearray):
+            self._hash.update(CodeHashPrefix.BYTEARRAY.value)
+            self._hash.update(obj)
+            self._hash.update(get_size_in_bytes(obj))
+
+        elif obj is None or isinstance(obj, (int, float, str, bool)):
+            if obj is None:
+                self._hash.update(CodeHashPrefix.NONE.value)
+            elif isinstance(obj, int):
+                self._hash.update(CodeHashPrefix.INT.value)
+            elif isinstance(obj, float):
+                self._hash.update(CodeHashPrefix.FLOAT.value)
+            elif isinstance(obj, str):
+                self._hash.update(CodeHashPrefix.STRING.value)
+            elif isinstance(obj, bool):
+                self._hash.update(CodeHashPrefix.BOOL.value)
+            json_bytes = json.dumps(obj).encode()
+            self._hash.update(json_bytes)
+            self._hash.update(get_size_in_bytes(json_bytes))
+
+        elif isinstance(obj, (list, tuple)):
+            if isinstance(obj, list):
+                self._hash.update(CodeHashPrefix.LIST.value)
+            elif isinstance(obj, tuple):
+                self._hash.update(CodeHashPrefix.TUPLE.value)
+            for elem in obj:
+                self._update_with_checks(elem, depth + 1)
+
+        elif isinstance(obj, dict):
+            self._hash.update(CodeHashPrefix.DICT.value)
+            for key, elem in obj.items():
+                self._update_with_checks(key, depth + 1)
+                self._update_with_checks(elem, depth + 1)
+
+        elif inspect.isroutine(obj):
+            self._hash.update(CodeHashPrefix.ROUTINE.value)
+            if obj.__defaults__ is not None:
+                for d in obj.__defaults__:
+                    self._update_with_checks(d, depth + 1)
+
+            self._analyze_code(obj.__code__, depth)
+
+        elif inspect.iscode(obj):
+            self._hash.update(CodeHashPrefix.CODE.value)
+            self._analyze_code(obj, depth)
+
+        else:
+            self._hash.update(CodeHashPrefix.DEFAULT.value)
+            message = oneline(
+                f"""
+                Found a constant {obj!r} of type {type(obj)!r} that
+                Bionic cannot hash. This is most likely a miss from the
```

Comment:
"a miss" and "the developers" are both vague. Let's just say "a bug in Bionic". Also, how should they contact us?

Finally, can we add a TODO to find out what the actual list of constant types is? I think this is a case where just doing the right thing (finding the list of constant types) is almost as easy as writing a clear explanation of the almost-right thing.